src/connector/src/source/cdc/external/postgres.rs

```diff
@@ -0,0 +1,332 @@
+        };
+
+        let txid: i64 = row.get::<_, i64>(0);
+        pg_offset.txid = txid;
+
+        // commit the transaction
+        trxn.commit().await?;
+
+        Ok(CdcOffset::Postgres(pg_offset))
+    }
+
+    fn parse_cdc_offset(&self, offset: &str) -> ConnectorResult<CdcOffset> {
+        Ok(CdcOffset::Postgres(PostgresOffset::parse_debezium_offset(
+            offset,
+        )?))
+    }
+
+    fn snapshot_read(
+        &self,
+        table_name: SchemaTableName,
+        start_pk: Option<OwnedRow>,
+        primary_keys: Vec<String>,
+    ) -> BoxStream<'_, ConnectorResult<OwnedRow>> {
+        self.snapshot_read_inner(table_name, start_pk, primary_keys)
+    }
+}
+
+impl PostgresExternalTableReader {
+    pub async fn new(
+        properties: HashMap<String, String>,
+        rw_schema: Schema,
+    ) -> ConnectorResult<Self> {
+        tracing::debug!(?rw_schema, "create postgres external table reader");
+
+        let config = serde_json::from_value::<ExternalTableConfig>(
+            serde_json::to_value(properties).unwrap(),
+        )
+        .map_err(|e| {
+            ConnectorError::Config(anyhow!(
+                "fail to extract postgres connector properties: {}",
+                e
+            ))
+        })?;
+
+        let database_url = format!(
+            "postgresql://{}:{}@{}:{}/{}",
+            config.username, config.password, config.host, config.port, config.database
+        );
+
+        let (client, connection) = tokio_postgres::connect(&database_url, NoTls).await?;
+
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                tracing::error!("connection error: {}", e);
+            }
+        });
+
+        let field_names = rw_schema
+            .fields
+            .iter()
+            .filter(|f| f.name != OFFSET_COLUMN_NAME)
+            .map(|f| format!("\"{}\"", f.name.as_str()))
+            .join(",");
+
+        Ok(Self {
+            config,
+            rw_schema,
+            field_names,
+            client: tokio::sync::Mutex::new(client),
+        })
+    }
+
+    #[try_stream(boxed, ok = OwnedRow, error = ConnectorError)]
+    async fn snapshot_read_inner(
+        &self,
+        table_name: SchemaTableName,
+        start_pk_row: Option<OwnedRow>,
+        primary_keys: Vec<String>,
+    ) {
+        let order_key = primary_keys.iter().join(",");
+        let sql = if start_pk_row.is_none() {
+            format!(
+                "SELECT {} FROM {} ORDER BY {}",
+                self.field_names,
+                self.get_normalized_table_name(&table_name),
+                order_key
+            )
+        } else {
+            let filter_expr = Self::filter_expression(&primary_keys);
+            format!(
+                "SELECT {} FROM {} WHERE {} ORDER BY {}",
+                self.field_names,
+                self.get_normalized_table_name(&table_name),
+                filter_expr,
+                order_key
+            )
```

Comment:
Should we parameterized column names in `SELECT` and `ORDER BY` as well as table name in `FROM` as well? I can think of a way to do SQL injection (but I won't share it here :)